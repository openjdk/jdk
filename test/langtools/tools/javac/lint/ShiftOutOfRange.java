/*
 * @test /nodynamiccopyright/
 * @bug 5038439
 * @summary Verify warnings about bit shifts using out-of-range shift amounts
 * @compile/ref=ShiftOutOfRange.out -XDrawDiagnostics -Xlint:lossy-conversions ShiftOutOfRange.java
 */

public class ShiftOutOfRange {

    public void shiftInt() {
        int a = 123;

        // These should generate warnings
        a = a << (byte)-1;
        a = a >> (char)-1;
        a = a >>> (short)-1;
        a <<= -1;
        a >>= -1L;          // also generates "implicit cast from long to int in compound assignment is possibly lossy"
        a >>>= (byte)-1;

        // These should not generate warnings
        a = a << (byte)0;
        a = a >> (char)7;
        a = a >>> (short)13;
        a <<= 17;
        a >>= (long)23;     // also generates "implicit cast from long to int in compound assignment is possibly lossy"
        a >>>= (byte)31;
        a <<= hashCode();
        a >>= hashCode();
        a >>>= hashCode();

        // These should generate warnings
        a = a << (byte)32;
        a = a >> (char)32;
        a = a >>> (short)32;
        a <<= 32;
        a >>= (long)32;     // also generates "implicit cast from long to int in compound assignment is possibly lossy"
        a >>>= (byte)32;
    }

    public void shiftLong() {
        long a = 123;

        // These should generate warnings
        a = a << (byte)-1;
        a = a >> (char)-1;
        a = a >>> (short)-1;
        a <<= -1;
        a >>= -1L;
        a >>>= (byte)-1;

        // These should not generate warnings
        a = a << (byte)0;
        a = a >> (char)15;
        a = a >>> (short)25;
        a <<= 34;
        a >>= (long)47;
        a >>>= (byte)63;
        a <<= hashCode();
        a >>= hashCode();
        a >>>= hashCode();

        // These should generate warnings
        a = a << (byte)64;
        a = a >> (char)64;
        a = a >>> (short)64;
        a <<= 64;
        a >>= (long)64;
        a >>>= (byte)64;
    }
}
