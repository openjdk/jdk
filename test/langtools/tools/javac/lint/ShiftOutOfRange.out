ShiftOutOfRange.java:14:18: compiler.warn.bit.shift.out.of.range: int, -1, 31
ShiftOutOfRange.java:15:18: compiler.warn.bit.shift.out.of.range: int, 65535, 31
ShiftOutOfRange.java:16:19: compiler.warn.bit.shift.out.of.range: int, -1, 31
ShiftOutOfRange.java:17:15: compiler.warn.bit.shift.out.of.range: int, -1, 31
ShiftOutOfRange.java:18:15: compiler.warn.possible.loss.of.precision: long, int
ShiftOutOfRange.java:18:15: compiler.warn.bit.shift.out.of.range: int, -1, 31
ShiftOutOfRange.java:19:16: compiler.warn.bit.shift.out.of.range: int, -1, 31
ShiftOutOfRange.java:26:15: compiler.warn.possible.loss.of.precision: long, int
ShiftOutOfRange.java:33:18: compiler.warn.bit.shift.out.of.range: int, 32, 0
ShiftOutOfRange.java:34:18: compiler.warn.bit.shift.out.of.range: int, 32, 0
ShiftOutOfRange.java:35:19: compiler.warn.bit.shift.out.of.range: int, 32, 0
ShiftOutOfRange.java:36:15: compiler.warn.bit.shift.out.of.range: int, 32, 0
ShiftOutOfRange.java:37:15: compiler.warn.possible.loss.of.precision: long, int
ShiftOutOfRange.java:37:15: compiler.warn.bit.shift.out.of.range: int, 32, 0
ShiftOutOfRange.java:38:16: compiler.warn.bit.shift.out.of.range: int, 32, 0
ShiftOutOfRange.java:45:18: compiler.warn.bit.shift.out.of.range: long, -1, 63
ShiftOutOfRange.java:46:18: compiler.warn.bit.shift.out.of.range: long, 65535, 63
ShiftOutOfRange.java:47:19: compiler.warn.bit.shift.out.of.range: long, -1, 63
ShiftOutOfRange.java:48:15: compiler.warn.bit.shift.out.of.range: long, -1, 63
ShiftOutOfRange.java:49:15: compiler.warn.bit.shift.out.of.range: long, -1, 63
ShiftOutOfRange.java:50:16: compiler.warn.bit.shift.out.of.range: long, -1, 63
ShiftOutOfRange.java:64:18: compiler.warn.bit.shift.out.of.range: long, 64, 0
ShiftOutOfRange.java:65:18: compiler.warn.bit.shift.out.of.range: long, 64, 0
ShiftOutOfRange.java:66:19: compiler.warn.bit.shift.out.of.range: long, 64, 0
ShiftOutOfRange.java:67:15: compiler.warn.bit.shift.out.of.range: long, 64, 0
ShiftOutOfRange.java:68:15: compiler.warn.bit.shift.out.of.range: long, 64, 0
ShiftOutOfRange.java:69:16: compiler.warn.bit.shift.out.of.range: long, 64, 0
27 warnings
